src/xserver/mod.rs

```diff
@@ -1,22 +1,229 @@
-use crate::control::{Controller, Eyes};
+use crate::control::{Controller, Eyes, ToBrain, ToController};
 
-pub struct XController;
-pub struct XEyes;
+use std::{ptr, sync::mpsc::{SyncSender, Receiver}, process::Command, time::Instant};
+use x11::xlib::{self, _XDisplay};
+use image::{ImageBuffer, Rgb};
+
+pub struct XController {
+    window: u64,
+}
+pub struct XEyes {
+    window: u64,
+}
+
+fn get_window_id_by_title(name: &str) -> eyre::Result<xlib::Window> {
+    let output = Command::new("xdotool")
+        .args(&["search", "--name", name])
+        .output()?;
+
+    let binding = String::from_utf8(output.stdout).expect("Error getting window id");
+    let window_id_str = binding.as_str();
+    let window_id = window_id_str.trim().parse()?;
+    Ok(window_id)
+}
+
+impl XEyes {
+    pub fn new(window_name: &str) -> eyre::Result<Self>{
+        Ok(Self {window: get_window_id_by_title(window_name)?})
+    }
+
+    
+    pub fn get_image(&self, display: *mut _XDisplay) -> eyre::Result<ImageBuffer<Rgb<u8>,Vec<u8>>> {
+        unsafe {
+            // Open the display
+        
+            // Get the window attributes
+            let mut window_attributes: xlib::XWindowAttributes = std::mem::zeroed();
+            xlib::XGetWindowAttributes(display,  self.window, &mut window_attributes);
+    
+            // Get the dimensions of the window
+            let width = window_attributes.width as u32;
+            let height = window_attributes.height as u32;
+            
+            // Create an XImage structure to hold the screenshot
+            let image = xlib::XGetImage(
+                display,
+                self.window,
+                0,
+                0,
+                width,
+                height,
+                xlib::XAllPlanes(),
+                xlib::ZPixmap,
+            );
+    
+            // Process the image data
+            let mut buffer: Vec<u8> = Vec::with_capacity((width * height * 4) as usize);
+    
+            for y in 0..height {
+                for x in 0..width {
+                    let pixel = xlib::XGetPixel(image, x as i32, y as i32);
+                    buffer.push(((pixel >> 16) & 0xFF) as u8); // Red
+                    buffer.push(((pixel >> 8) & 0xFF) as u8);  // Green
+                    buffer.push((pixel & 0xFF) as u8);         // Blue
+                }
+            }
+    
+            // Create an ImageBuffer from the pixel data
+            let image_buffer = ImageBuffer::<Rgb<u8>, _>::from_raw(width, height, buffer).unwrap();
```

Comment:
I would prefer this to be running on another thread, but it's not a dealbreaker. Seems it would be pretty difficult to do with the way the code is structured.